Library/Homebrew/utils/lock.sh

```diff
@@ -0,0 +1,30 @@
+HOMEBREW_LOCK_DIR="$HOMEBREW_LIBRARY/lock.d"
+HOMEBREW_LOCK_PID_FILE="$HOMEBREW_LOCK_DIR/PID"
+
+lock() {
+  local pid
+
+  if mkdir "$HOMEBREW_LOCK_DIR" &>/dev/null
+  then
+    echo "$$" > "$HOMEBREW_LOCK_PID_FILE"
+  else
+    # lock failed, check if the other PID is alive
+    pid="$(<"$HOMEBREW_LOCK_PID_FILE")"
+
+    if [[ $? = 0 ]] && ! kill -0 "$pid" &>/dev/null
+    then
+      # lock is stale, remove it and restart
+      unlock
```

Comment:
I'm not sure this will behave 100% correctly. I believe it could be stealing the lock from another process, but I have to think some more about this (and maybe run some tests). But I'm happy to hear about why I'm mistaken.